metals/src/main/scala/scala/meta/internal/metals/callHierarchy/CallHierarchyItemBuilder.scala

```diff
@@ -0,0 +1,161 @@
+package scala.meta.internal.metals.callHierarchy
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import scala.meta.internal.metals.BuildTargets
+import scala.meta.internal.metals.Compilers
+import scala.meta.internal.metals.HoverExtParams
+import scala.meta.internal.metals.Icons
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.semanticdb.Scala._
+import scala.meta.internal.semanticdb.SymbolOccurrence
+import scala.meta.internal.semanticdb.TextDocument
+import scala.meta.internal.{semanticdb => s}
+import scala.meta.io.AbsolutePath
+import scala.meta.pc.CancelToken
+
+import org.eclipse.{lsp4j => l}
+
+private[callHierarchy] final class CallHierarchyItemBuilder(
+    workspace: AbsolutePath,
+    icons: Icons,
+    compilers: () => Compilers,
+    buildTargets: BuildTargets,
+) {
+  private def buildItem(
+      name: String,
+      kind: l.SymbolKind,
+      uri: String,
+      range: l.Range,
+      selectionRange: l.Range,
+      detail: String,
+      data: CallHierarchyItemInfo,
+  ): l.CallHierarchyItem = {
+    val item = new l.CallHierarchyItem(name, kind, uri, range, selectionRange)
+    item.setDetail(detail)
+    item.setData(data)
+    item
+  }
+
+  // When we search the call hierarchy of a class, we also need to search the symbol of the primary constructor
+  private def getContructorRelatedSymbol(
+      info: s.SymbolInformation
+  ): Array[String] =
+    info.signature match {
+      case s.ClassSignature(_, _, _, declarations) =>
+        val constructor = declarations
+          .flatMap(
+            _.symlinks.find(_.endsWith("`<init>`()."))
+          )
+        // for case class apply is like a constructor
+        val applySymbol = Option.when(info.isCase)(
+          info.symbol.replace("#", ".apply().")
```

Comment:
What about user defined apply? The symbol might be different, so you would need to search for all apply :thinking:

Reply:
The apply method of case class is special, because when we click on the name of the class, we want to take in consideration the constructor. For example:

```scala
case class Fo@@o(x: Int)
...
<<Foo>>(0)
```
`.apply().` is a special symbol in this case that appear in synthetics. The feature work well with user defined feature, I just add a test for it.

Reply:
ok cool!